enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md

```diff
@@ -0,0 +1,306 @@
+
+3. See [this proposal](https://github.com/openshift/enhancements/pull/214) for additional user scenarios that could 
+utilize this proposal.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Install
+
+A comparison of installing the implementation of this proposal via OLM vs. CVO was done.  The net result was 
+to go with CVO.
+
+Some specifics:
+- During the review for the ["Installation of CSI drivers that replace in-tree drivers](https://github.com/openshift/enhancements/blob/master/enhancements/storage/csi-driver-install.md)
+a question was posed about whether the exact type of CSI driver this proposal is describing would use the method 
+described in that document (namely OLM).  The recommendation from the [storage SMEs was to go with CVO](https://github.com/openshift/enhancements/pull/139/files#r391761362)
+- In particular, using CVO and having it be a day 1 install operation meets the ease of use targets we are trying to
+reach with this proposal
+- Crafting dependencies between this proposal and [related proposals](https://github.com/openshift/enhancements/pull/214)
+at the time of this writing is immature.
+
+#### Brief overview of CSI and the main K8S hook points
+
+ - a [CSI plugin](https://github.com/container-storage-interface/spec) is a gRPC endpoint that implements CSI services and
+ allows storage vendors to develop a storage solution that works across a number of container orchestration systems.
+ - though for our purposes, we will not provide a generic CSI plugin, but rather a special case one that is only intended
+ to work on OpenShift
+ - also for our purposes, `plugin`, `driver`, and `endpoint` seem to be the same thing based on which document you are
+ reading.
+ - k8s provides Kubernetes CSI Sidecar Containers: a set of standard containers that aim to simplify the development and 
+ deployment of CSI Drivers on Kubernetes.
+ - There are CSI volume and persistent volume types in k8s.
+ - per [k8s volume docs](https://kubernetes.io/docs/concepts/storage/volumes/#csi) "the CSI types do not support direct
+ reference from a Pod and may only be referenced in a Pod via a `PersistentVolumeClaim` object" (direct quote)
+ - However, @gnufied tells us that documentation is out of date and incorrect.  To quote him:  That is where [inline 
+ CSI volumes come](https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html)
+ - To that end, one of the sidecar containers, the [external provisioner](https://kubernetes-csi.github.io/docs/external-provisioner.html),
+ helps with this
+ - It watches Kubernetes `PersistentVolumeClaim` objects, and if the PVC 
+ references a Kubernetes StorageClass, and the name in the provisioner field of the storage class matches the name 
+ returned by the specified CSI endpoint `GetPluginInfo` call, it calls the CSI endpoint `CreateVolume` method to provision
+ the new volume.
+ - there are also lifecycle methods for destroy, resize, etc.
```

Comment:
> The volume can be probably injected by an admission plugin easily, I am not sure how to inject volumeMount, especially the mountPath.

We plan on having a separate admission controller that injects the volumeMount with the correct mountPath. This is out of scope for the proposal here.

Reply:
> The main benefit of ephemeral CSI volume is that you don't need a PVC in user's namespace (who's going to delete it and when?) and you don't need to implement Create/DeleteVolume and Stage/UnstageVolume in your driver.

This is a _huge_ benefit IMO. Is the goal for this feature to be GA by 4.6? Or alternatively, would we be comfortable shipping a CSI driver like this in 4.6 that utilizes the current beta API?

Reply:
> would we be comfortable shipping a CSI driver like this in 4.6 that utilizes the current beta API?

Yes, if the CSI driver is marked as Technical Preview. Which I would suggest for a first public release of any CSI driver anyway.

Reply:
> I would strongly recommend using ephemeral CSI volumes, https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html

Yep @gnufied mentioned this briefly as well last week in https://github.com/openshift/enhancements/pull/250#discussion_r394562356 

> 
> It will be officially technical preview in OCP 4.5, but it's already beta in Kubernetes 1.17 and thus enabled by default in OCP 4.4. It's only not documented in OCP docs.
> 
> In that sense:
> 
> ```
> apiVersion: v1
> kind: Pod
> metadata:
>   name: some-pod
> spec:
>   containers:
>     ...
>      volumeMounts:
>         - name: entitlement
>           mountPath: /etc/entitlement
>   volumes:
>       - name: entitlement
>         csi:
>           driver: secretprovider.openshift.io
>           volumeAttributes:
>               namespace: openshift-config
>               name: entitlement
> ```
> 
> The `volume` can be probably injected by an admission plugin easily, I am not sure how to inject `volumeMount`, especially the `mountPath`.
> 
> The CSI driver then gets only `NodePublish` and `NodeUnpublish` call for each Pod that uses the volume. `NodePublish` does whatever you need - it could mount tmpfs and put the Secret data there and refresh it when it changes on API server. Or it could bind-mount a driver's directory (e.g. `/var/lib/secretprovider.openshift.io/<namespace>/<name>`) with the data, so you need to refresh it at one place for all pods on the node. With bind-mount, you must ensure that any Pod that needs the data can actually read it - for example have proper SELinux context and fs permissions. Read-only bind mount is strongly suggested.
> 
> The main benefit of ephemeral CSI volume is that you don't need a PVC in user's namespace (who's going to delete it and when?) and you don't need to implement Create/DeleteVolume and Stage/UnstageVolume in your driver.

Very very helpful drill down ... thanks

> 
> I'd recommend you to look at https://github.com/kubernetes-sigs/secrets-store-csi-driver - it implements a Secrets-like CSI driver that gets the secret data via plain `exec()`. It does not seem to implement refreshing of the data though.

Interestingly, that example was referenced in an initial conversation between @derekwaynecarr @bparees @gnufied and myself a few weeks ago but was subsequently ruled out.

Qutotes from @gnufied 
- https://coreos.slack.com/archives/GU5A8GSDR/p1582749384007000
- https://coreos.slack.com/archives/GU5A8GSDR/p1582749408007300

Still mulling as I type, and I still need to do a proper read of https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html, but at first blush, this feels like needing to specify both approaches as possible alternatives, at least in the interm, with pros/cons of each.

Reply:
So completed my latest dive into https://github.com/kubernetes-sigs/secrets-store-csi-driver to better understand it. 

I tried to stand it up end to end like I did with csi-hostpath using its set of try it out steps.  I was unsuccessful, as using the 2 providers (azure and hashicorp) they cite did not work in my 4.x cluster (hostpath permissions, can't create directories, etc.) that I'm guessing they do not hit on vanilla k8s.  Or they have bugs.

In the end, I don't think it matters.  With more iterations and all the new info in this thread, I can piece together a roadmap using this style CSI plugin.

1) https://github.com/Azure/secrets-store-csi-driver-provider-azure/blob/master/pkg/azure/provider.go was a useful enough example of a storage class provider

2) the csi code at https://github.com/openshift/origin/tree/master/vendor/k8s.io/kubernetes/pkg/volume/csi makes more sense to me now

3) including https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/volume/csi/csi_mounter.go#L105-L254 which handles `NodePublishVolume` for both volsrc's and pvs, and maps both the VolumeAttributes @jsafrane noted in his Pod example, as well as vol src same namespace secret refs (https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/staging/src/k8s.io/api/core/v1/types.go#L1736) and pv/pvc any namesapce secret refs https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/staging/src/k8s.io/api/core/v1/types.go#L1674-L1705

4) https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/volume/csi/csi_util.go#L44-L55 for converting corev1.Secrets to the map[string]string signatures for secrets you see in the CSI APIs

With all that, I'll start articulating an ephemeral CSI volume option in addition to the existing PVC based one.

As previously noted, not needing to create a PVC is a big plus for ephemeral volumes.
Multi namespace secret refs seem like it could be a plus for the `CSIPersistentVolumeSource` (in conjunction with PVC) option, but seems like @jsafrane 's idea around using volumeAttributes to pass that metadata is an acceptable workaround.